### PR TITLE
Embed new nsxiv icon

### LINKED ISSUE #104: Embed the new icon in nsxiv
Currently, `data.h` uses the old icon. We should either build a new `data.h` or rework the icon system.

See #1 and #26 for the previous discussions.

### PR DESCRIPTION
**Pull request by mamg22**
This adds a way to automatically embed the application icons into the `nsxiv` binary. The `img2data` program takes the icon filenames as arguments and outputs C arrays for each one, adding the color palette and icon list arrays at the end. The generated output is put into a file later included by `data.h` when building.

Closes #104

<details>
 <summary>About the format muennich used for the icons:</summary>
The data in each icon data array on `data.h` is compressed using run-length encoding, where the four high bits of each `char` are the run length (minus one) and the lowest four bits are the data, which are indexes for `icon_colors[]`.

The `dat2h.awk` did the run-length encoding of the image data. The format it expected was:

```
#color
#color
...
#color

#color
#color
...
```

where each `#color` is the ARGB color for each pixel, left-to-right and then top-to-bottom, each image's pixel color list was separated by an empty line. At the end, it printed the colors gathered from all images to be stored in `icon_colors[]`.

What they used for generating the input file is still unknown to me (maybe ImageMagick?).
</details>

### CODE REVIEW

File: icon/img2data.c [hunk 1]
```diff
@@ -0,0 +1,152 @@
+#include <stdio.h>
+#include <stdlib.h>
+
+#include <Imlib2.h>
+
+void die(const char *message)
```
Comment (N-R-K) [inline icon/img2data.c:6]: think you should be able to use `error` function from util.c here.

**Comment by N-R-K** [comment]
Haven't tested or taken a proper look but here's a couple things:
Should we just hardcode the thing? We can leave the source in the repo for anyone that wants to use some other icon, but don't think we need to touch the Makefile at all and just hardcoding it should be fine. @eylles

Also the source uses a good ton of c99 features. Since this isn't part of the application (assuming we take the hardcode path) it's not a big problem. But if we want to have it be automatic and add it to the build then we should make it c89 compliant since the rest of the codebase is c89.

**Comment by XPhyro** [comment]
I don't see ourselves changing the icon frequently enough (if ever) for this to be worth automated, or worth extra processing on the user. Just as sxiv, hardcoding should be fine.

**Comment by eylles** [comment]
I'm also on the keeping it hardcoded approach but we should keep source, perhaps add a secondary readme inside the icon dir that is a copy of @mamg22 's opening comment for the sake of documentation on the format of the icons would be a good idea.

**Comment by XPhyro** [comment]
The output of img2data.c looks like we're embedding only one icon, unlike the old data.h where sizes of 16, 32, 48, 64 and 128 were embedded. @mamg22 Are we embedding only one size?

Also, we should explain how to compile this program in the readme. Not all might figure out linking to Imlib2 or nsxiv's codebase (util.c and nsxiv.h).

Lastly, I'd rather if this didn't link with nsxiv. It would be easier to incorporate into other open-source programs.

**Comment by N-R-K** [comment]
> Lastly, I'd rather if this didn't link with nsxiv. It would be easier to incorporate into other open-source programs.

Currently it only uses the `error` function which is about 10 lines. It should be trivial to copy paste it from util.c if one wanted.

**Comment by N-R-K** [comment]
@XPhyro removed dependency on util.c.

File: icon/data.h [hunk 2]
```diff
@@ -1,247 +1,17 @@
 #ifndef ICON_DATA_H
 #define ICON_DATA_H
 
+#include "../nsxiv.h"
```
Comment (N-R-K) [inline icon/data.h:4]: Why is this needed?

File: icon/img2data.c [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+}
+
+int main(int argc, char **argv)
+{
+	unsigned int img_size = 0;
+	unsigned int i;
+
+	if (argc < 2)
+		error(EXIT_FAILURE, 0, "No icons provided");
+	else if (argc > 1 + ARRLEN(icon_sizes))
```
Comment (N-R-K) [inline icon/img2data.c:145]: Might be better to do this instead.
```suggestion
	else if (argc-1 > ARRLEN(icon_sizes))
```

File: icon/data.h (hunk 2 above)
Comment (N-R-K) [reply icon/data.h]: Compiles fine without, so removed it.

**Review by N-R-K — APPROVED** [review]
I haven't tested this to see if the embedded icon is correct or not. Assuming you guys have @eylles @XPhyro , I'll approve this.

**Review by XPhyro — APPROVED** [review]
`rofi -show window -show-icons -window-thumbnail` shows the icon correctly. I assume this would be the case in a WM/DE too.

**Comment by N-R-K** [comment]
Hmm, `palette` should probably be `unsigned long` as well.

**Comment by N-R-K** [comment]
@XPhyro can you test 561706e out? If it's working fine, then I think we're good to merge.

**Comment by XPhyro** [comment]
> @XPhyro can you test [561706e](https://github.com/nsxiv/nsxiv/commit/561706eb76278c92fe9ee31cf29706ce452bc035) out? If it's working fine, then I think we're good to merge.

Still works correctly.

File: icon/data.h (hunk 2 above)
Comment (mamg22) [reply icon/data.h]: > Why is this needed?

That was just to silence some LSP errors, but yeah, it compiles fine without it.
